### PR TITLE
Add ERC721 hooks

### LINKED ISSUE #965: Add Hooks to ERC721 and ERC1155
We added Hooks to ERC20 for integrating the ERC20Votes extension, and they should be added to other tokens for consistency since it could be helpful as well.

### PR DESCRIPTION
**Pull request by ericnordelo**
Partially fixes #965  

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Added entry to CHANGELOG.md 
- [ ] Tried the feature on a public network

### CODE REVIEW

File: src/token/erc721/erc721.cairo
Comment (ericnordelo) [reply src/token/erc721/erc721.cairo:364]: The public method still panics, this is an internal one, I don't feel like the warning is necessary. Maybe a changelog-specific entry, but we would need to add a few, and since we are breaking that often, I think we don't need to be that specific with the internal API.

File: src/token/erc721/erc721.cairo
Comment (ericnordelo) [reply src/token/erc721/erc721.cairo:381]: I think there's no harm in keeping it. We have other examples of internal utilities that are not used directly in the component (like _safe_mint). It makes it easier for users to know when the token "exists".

File: src/token/erc721/erc721.cairo
Comment (ericnordelo) [reply src/token/erc721/erc721.cairo:454]: Since users can mint to contracts unable to handle the token when no using safe_mint, I think it should. It is also consistent with the Solidity version.

File: src/token/erc721/erc721.cairo
Comment (ericnordelo) [reply src/token/erc721/erc721.cairo:458]: mmm, I think it was for consistency, but now I don't see why. Rolling it back.

File: src/tests/token/test_erc721.cairo
Comment (ericnordelo) [reply src/tests/token/test_erc721.cairo:752]: If it is called from the zero address, the auth is the zero address, which means there's no authorization check in the update. This is not possible in a real scenario (for the method we are testing here). 

The first check in `_update` is the token id check, but only if the auth address is passed. We don't want to check that a token has a previous owner, when the token is minted.

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/erc721.adoc:1]: Good catch! Done.

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/erc721.adoc:527]: After giving it some time, I think we can remove the requirement.

Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/erc721.adoc:527]: And we don't need the check for the operator != zero either, since the zero address is always unauthorized in the check_authorized method.

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/erc721.adoc:537]: That's a very good point, I will update it.

Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/erc721.adoc:537]: Done! Let me know what you think.

File: src/token/erc721/erc721.cairo
Comment (andrew-fleming) [reply src/token/erc721/erc721.cairo:363]: > The public method still panics, this is an internal one, I don't feel like the warning is necessary.

My concern is the internal fn panicked in this lib for so long, but yeah. It _is_ internal and we have everything documented in the API doc so that should be enough

> Maybe a changelog-specific entry, but we would need to add a few, and since we are breaking that often, I think we don't need to be that specific with the internal API.

Agreed, it's probably not necessary to be that specific especially in this case

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/api/erc721.adoc]: Right, but this would allow `is_approved_for_all` to potentially return true when the operator is zero, no?

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/api/erc721.adoc]: Looks much better :)

**Review by andrew-fleming — COMMENTED** [review]
Changes look good! I think we should also add tests for the new internals (`_require_owned`, `_check_authorized`, `_update`). And we should also test the behavior of hooks. Since ERC20 should have hooks tests as well, maybe we can create a separate issue/PR to handle hooks testing (and not further block this PR). WDYT?

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/erc721.adoc]: Yes, but I don't think there's a real issue with that. Even if people approve the zero address, if for some reason the zero address approval causes issues, there must be a big bug somewhere in the process, right?

**Comment by ericnordelo** [comment]
I added the tests for the internal methods that were missing, but not sure what's the best approach to test hooks, since the only implementation we provide is empty, and how can we test an empty behavior? If we were to test that functions don't do things that's infinite possibilities right? Any ideas?

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/api/erc721.adoc]: > Even if people approve the zero address, if for some reason the zero address approval causes issues, there must be a big bug somewhere in the process, right?

Totally agree. My uneasiness with removing the check is the fact that it allows the getter to return an incorrect value. To me, that's a bug (a very small one in this context but a bug nonetheless). Do you disagree?

**Comment by andrew-fleming** [comment]
> not sure what's the best approach to test hooks, since the only implementation we provide is empty, and how can we test an empty behavior? If we were to test that functions don't do things that's infinite possibilities right? Any ideas?

We must test ALL possibilities!!! 

No, there's no point in testing empty behaviors. What we can test is the side effects of the before and after hook with a mock to confirm they're called in the correct order i.e. `before_update` is called before `_update`. A Cairo mocking library would be super useful :(

Unless you have another idea, I say let's get this merged and regroup on this. WDYT?

**Review by andrew-fleming — COMMENTED** [review]
I left a couple tiny suggestions on the new tests. We just need to conclude on the open discussion and this should be good to go!

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [reply docs/modules/ROOT/pages/api/erc721.adoc]: Fair point. I think keeping it as consistent as possible with Solidity is good anyway. Let's keep the operator not zero check.

File: src/tests/token/test_erc721.cairo
Comment (ericnordelo) [reply src/tests/token/test_erc721.cairo:1338]: I do recognize the name is quite confusing, but not really. If there's no auth, and no event needs to be emitted, this method doesn't check that the token exists. The intent of this is to be used like this only inside _update. Maybe a comment is helpful in this context. Adding it.

File: src/tests/token/test_erc721.cairo
Comment (ericnordelo) [reply src/tests/token/test_erc721.cairo:1439]: Good catch!

File: src/tests/token/test_erc721.cairo
Comment (andrew-fleming) [reply src/tests/token/test_erc721.cairo:1352]: Ahh my bad, I get the idea

**Review by andrew-fleming — APPROVED** [review]
Last suggestion: let's add the non-zero `operator` requirement for `_set_approval_for_all` in the API doc and then this should be good to go!

**Pull request merged**
